File: backend/microservice/src/level/objects/level/LevelData.scala
package microservice.level.objects.level

import io.circe.generic.semiauto._
import microservice.level.objects.inventory.{BirdInventory, BirdPool}
import microservice.level.objects.terrain._
import io.circe.{Decoder, Encoder}

/** 关卡编辑器 JSON 根结构，嵌入 Level.data 字段。
  *
  * 字段说明：
  *   - world：物理世界尺寸与重力
  *   - ground / terrain：地面与地形边界（可选，旧格式兼容 ground）
  *   - birdInventory：玩家可用鸟数量
  *   - obstacles / enemies：关卡内障碍物与敌人列表
  *   - backgroundTemplateId：背景模板 ID（可选）
  *   - birdPool：Director 配置的鸟池限制（可选）
  */
final case class LevelData(
  world: GameWorld,
  ground: Option[LevelGround],
  terrain: Option[LevelTerrain],
  birdInventory: BirdInventory,
  obstacles: List[LevelObstacle],
  enemies: List[LevelEnemy],
  backgroundTemplateId: Option[String] = None,
  birdPool: Option[BirdPool] = None
)

object LevelData {
  implicit val encoder: Encoder[LevelData] = deriveEncoder
  implicit val decoder: Decoder[LevelData] = deriveDecoder
}

File: frontend/src/objects/level/level/level-data.ts
import { z } from "zod";
import { nullishToUndefined } from "../../system/schema-utils.js";
import { BirdInventorySchema } from "../inventory/bird-inventory.js";
import { BirdPoolSchema } from "../inventory/bird-pool.js";
import { GameWorldSchema } from "./game-world.js";
import { LevelGroundSchema } from "../terrain/level-ground.js";
import { LevelTerrainSchema } from "../terrain/level-terrain.js";
import { LevelEnemySchema } from "../terrain/level-enemy.js";
import { LevelObstacleSchema } from "../terrain/level-obstacle.js";

export const LevelDataSchema = z.object({
  world: GameWorldSchema,
  ground: nullishToUndefined(LevelGroundSchema),
  terrain: nullishToUndefined(LevelTerrainSchema),
  birdInventory: BirdInventorySchema,
  birdPool: nullishToUndefined(BirdPoolSchema),
  obstacles: z.array(LevelObstacleSchema),
  enemies: z.array(LevelEnemySchema),
  backgroundTemplateId: nullishToUndefined(z.string().min(1)),
});

export type LevelData = z.infer<typeof LevelDataSchema>;
export { BirdInventorySchema } from "../inventory/bird-inventory.js";
export type { BirdInventory } from "../inventory/bird-inventory.js";
export { GameWorldSchema } from "./game-world.js";
export { LevelEnemySchema } from "../terrain/level-enemy.js";
export { LevelObstacleSchema } from "../terrain/level-obstacle.js";
export { BirdPoolSchema } from "../inventory/bird-pool.js";
export type { BirdPool } from "../inventory/bird-pool.js";
export type { GameWorld } from "./game-world.js";
export type { LevelEnemy } from "../terrain/level-enemy.js";
export type { LevelObstacle } from "../terrain/level-obstacle.js";
